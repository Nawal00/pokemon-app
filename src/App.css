* {
  box-sizing: border-box;
}

.App > h1 {
  text-align: center;
  padding: 60px;
  margin: 0 0 5rem 0;
  color: #ffffff;
  background-color: #4865c2;
}

.container {
  margin: auto;
  display: flex;
  width: 90%;
}

.info-message {
  background-color: #4865c2;
  height: fit-content;
  width: fit-content;
  padding: 0 0.5rem;
  color: #ffffff;
  border-radius: 5px;
}

.pokemon-list-wrapper {
  flex-basis: 50%;
  display: grid;
  grid-gap: 2rem;
  grid-template-columns: 50% 50%;
  overflow-y: auto;
  max-height: 100vh;
}
