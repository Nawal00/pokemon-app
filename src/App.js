import { useState, useEffect } from "react";
import PokemonList from "./components/PokemonList";
import PokemonCard from "./components/PokemonCard";
import "./App.css";

function App() {
  const [pokemonInfo, setPokemonInfo] = useState({});
  const [pokemonList, setPokemonList] = useState([]);
  const [loading, setLoading] = useState(false);
  const hasPokemonInfoData = Object.keys(pokemonInfo).length > 0;

  const fetchPokemons = async () => {
    const generationId = 1;
    const url = `https://pokeapi.co/api/v2/generation/${generationId}`;

    try {
      const response = await fetch(url);
      const { pokemon_species } = await response.json();

      setPokemonList(pokemon_species);
    } catch (error) {
      console.log(error);
    }
  };

  const fetchPokemonInfo = async (name) => {
    const url = `https://pokeapi.co/api/v2/pokemon/${name}/`;

    try {
      setLoading(true);
      const response = await fetch(url);
      const data = await response.json();

      setPokemonInfo(data);
    } catch (error) {
      console.log(error);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchPokemons();
  }, []);

  return (
    <div className="App">
      <h1>Generation One Pokemons</h1>
      <div className="container">
        <div className="pokemon-list-wrapper">
          {pokemonList?.map(({ name }) => (
            <PokemonList
              key={name}
              name={name}
              fetchPokemonInfo={fetchPokemonInfo}
            />
          ))}
        </div>
        {hasPokemonInfoData ? (
          <PokemonCard
            {...pokemonInfo}
            loading={loading}
            setPokemonInfo={setPokemonInfo}
          />
        ) : (
          <div className="info-message">
            <h3>Click on the pokemon name to view pokemon info</h3>
          </div>
        )}
      </div>
    </div>
  );
}

export default App;
