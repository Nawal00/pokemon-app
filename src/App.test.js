import { render, screen, act } from "@testing-library/react";
import App from "./App";

jest.spyOn(global, "fetch");
const generationId = 1;

describe("Given the `App` component", () => {
  beforeEach(async () => {
    fetch.mockResolvedValue({
      json: () =>
        Promise.resolve({
          pokemon_species: [{ name: "pikachu" }, { name: "ash" }],
        }),
    });
  });

  it("should render correctly", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("should request the correct url", async () => {
    await act(async () => render(<App />));

    expect(fetch).toHaveBeenCalledWith(
      `https://pokeapi.co/api/v2/generation/${generationId}`
    );
  });

  it("should have pokemon list rendered", async () => {
    await act(async () => render(<App />));

    expect(screen.getByText("pikachu")).toBeInTheDocument();
    expect(screen.getByText("ash")).toBeInTheDocument();
  });
});
