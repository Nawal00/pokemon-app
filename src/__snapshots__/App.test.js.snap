// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the \`App\` component should render successfully 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Generation One Pokemons
    </h1>
    <div
      class="container"
    >
      <div
        class="pokemon-list-wrapper"
      />
      <div
        class="info-message"
      >
        <h3>
          Click on the pokemon name to view pokemon info
        </h3>
      </div>
    </div>
  </div>
</div>
`;
